packages/block-editor/src/store/reducer.js

```diff
@@ -755,6 +755,31 @@ const withResetControlledBlocks = ( reducer ) => ( state, action ) => {
 	return reducer( state, action );
 };
 
+function cloneDeep( object ) {
+	return ! object ? {} : JSON.parse( JSON.stringify( object ) );
+}
+
+const addPatternCategoryToMetadata = ( reducer ) => ( state, action ) => {
+	if ( action.type === 'INSERT_BLOCKS' ) {
+		const { blocks, meta } = action;
+		const nextState = reducer( state, action );
+		const firstBlockClientId = blocks[ 0 ].clientId;
+		const firstBlockAttributes = cloneDeep( blocks[ 0 ].attributes );
+		if ( meta?.category && blocks.length === 1 ) {
+			nextState.attributes.set( firstBlockClientId, {
+				...firstBlockAttributes,
+				metadata: {
+					...( firstBlockAttributes.metadata || {} ),
+					categories: [ meta.category.name ],
+				},
+			} );
+		}
+		return nextState;
```

Comment:
Either this isn't being returned correctly, or the change is being overwritten later...

Reply:
Yeah, you still need to be calling the reducer here.